tensorflow/core/kernels/mkl_conv_ops.cc

```diff
@@ -42,26 +43,287 @@ limitations under the License.
+             memory::dims dilations,
+             memory::dims padding_l,
+             memory::dims padding_r) {
+    strides_ = strides;
+    padding_l_ = padding_l;
+    padding_r_ = padding_r;
+
+    // create memory descriptors for convolution data w/ no specified format
+    src_md_.reset(new memory::desc({src_dims}, MklDnnType<T>(),
+                                   memory::format::any));
+
+    filter_md_.reset(new memory::desc({filter_dims},
+                  MklDnnType<T>(), memory::format::any));
+
+    dst_md_.reset(new memory::desc({dst_dims}, MklDnnType<T>(),
+                                   memory::format::any));
+    if (!bias_dims.empty())
+        bias_md_.reset(new memory::desc({bias_dims}, MklDnnType<T>(),
+                                   memory::format::any));
+    // create a convolution
+    if (!bias_dims.empty()) {
+      fwd_desc_.reset(new convolution_forward::desc(prop_kind::forward,
+                      convolution_direct, *src_md_, *filter_md_,
+                      *bias_md_, *dst_md_, strides_, dilations,
+                      padding_l_, padding_r_, padding_kind::zero));
+    } else {
+      fwd_desc_.reset(new convolution_forward::desc(prop_kind::forward,
+                      convolution_direct, *src_md_, *filter_md_,
+                      *dst_md_, strides_, dilations,
+                      padding_l_, padding_r_, padding_kind::zero));
+    }
+
+    fwd_pd_.reset(new convolution_forward::primitive_desc(
+                  *fwd_desc_, cpu_engine_));
+
+    // store the expected memory format
+    src_fmt_ = static_cast<mkldnn::memory::format>(
+                 fwd_pd_.get()->src_primitive_desc().desc().data.format);
+
+    filter_fmt_ = static_cast<mkldnn::memory::format>(
+                 fwd_pd_.get()->weights_primitive_desc().desc().data.format);
+    dst_fmt_ = static_cast<mkldnn::memory::format>(
+                 fwd_pd_.get()->dst_primitive_desc().desc().data.format);
+
+    // create memory primitive based on dummy data
+    src_mem_.reset(new memory(fwd_pd_.get()->src_primitive_desc(), DummyData));
+    filter_mem_.reset(new memory(fwd_pd_.get()->weights_primitive_desc(),
+                      DummyData));
+    dst_mem_.reset(new memory(fwd_pd_.get()->dst_primitive_desc(), DummyData));
+
+    // create convolution primitive and add it to net
+    if (!bias_dims.empty()) {
+        bias_mem_.reset(new memory({{{bias_dims}, MklDnnType<T>(),
+                             memory::format::x}, cpu_engine_}, DummyData));
+        conv_fwd_.reset(new convolution_forward(*fwd_pd_, *src_mem_,
+                        *filter_mem_, *bias_mem_, *dst_mem_));
+    } else {
+        conv_fwd_.reset(new convolution_forward(*fwd_pd_, *src_mem_,
+                        *filter_mem_, *dst_mem_));
+    }
+
+    fwd_primitives_.push_back(*conv_fwd_);
+    return;
+  }
+
+ public:
+  // Convolution forward execute with bias
+  void Execute(void* src, void* w, void* b, void* dst) {
```

Comment:
Please template this function on the input types instead of using void*.  If set_data_handle only accepts void* then we want to cast as late as possible, i.e. in the call to set_data_handle itself.

Reply:
Actually MKL DNN C/C++ interface is as follows
  inline void set_data_handle(void *handle) const;      // mkldnn.hpp -- open source API
so passing "void*"  buffer pointers is the natural way to do it.

Reply:
If stressed, we can template "Execute" and then cast input parameters when invoking 
set_data_handle.

Reply:
That's what I was proposing.

Reply:
Changed per your suggestion